csrc/ops/composite.cpp

```diff
@@ -299,5 +299,73 @@ TensorView* view_as_real(TensorView* x) {
   auto tv_vector = bitCastOp(vec_type, x);
   return viewAsScalar(tv_vector);
 }
+namespace {
+
+//! Create new output for matmul
+static TensorView* newForMatmul(TensorView* tv_a, TensorView* tv_b) {
+  auto orig_domain_a =
+      TensorDomain::noReductions(tv_a->getMaybeRFactorDomain());
+  auto orig_domain_b =
+      TensorDomain::noReductions(tv_b->getMaybeRFactorDomain());
+
+  auto ndims_a = orig_domain_a.size();
+  auto ndims_b = orig_domain_b.size();
+
+  NVF_ERROR(ndims_a >= 1 && ndims_b >= 1);
+
+  // Matmul output size is same as the higher dimensional input size if both A/B
+  // > 1D.
+  auto ndims_out = std::max(ndims_a, ndims_b);
+  if (std::min(ndims_a, ndims_b) == 1) {
+    // If one of the inputs is 1D, the output size is 1 less than the higher
+    // dimensional input size, since either M/N axis will be missing in the
+    // output. For eg: [M, K] x [K] -> [M]
+    ndims_out = std::max(ndims_a, ndims_b) - 1;
+  }
+
+  std::vector<IterDomain*> out_domain(ndims_out, nullptr);
+
+  const std::vector<IterDomain*>& mapping_a = ops::mapMatmulOpIterDomains(
+      orig_domain_a, MatmulRole::INPUT_A, ndims_out);
+  const std::vector<IterDomain*>& mapping_b = ops::mapMatmulOpIterDomains(
+      orig_domain_b, MatmulRole::INPUT_B, ndims_out);
+
+  for (auto idx : c10::irange(ndims_out)) {
+    std::vector<IterDomain*> input_ids;
+    input_ids.reserve(2);
+    if (mapping_a[idx] != nullptr) {
+      input_ids.emplace_back(mapping_a[idx]);
+    }
+    if (mapping_b[idx] != nullptr) {
+      input_ids.emplace_back(mapping_b[idx]);
+    }
+    out_domain[idx] = ops::newOutputIterDomain(input_ids);
+  }
+
+  TensorDomain* td = IrBuilder::create<TensorDomain>(
+      out_domain, TensorDomain::getContiguityFilledWith(out_domain, true));
+
+  return IrBuilder::create<TensorView>(td, tv_a->dtype());
+}
+
+} // namespace
+
+// TODO (Priya): This will be renamed to matmul once we are ready to modify the
+// python API backend. Keeping separate for now, to avoid breaking tests in
+// Thunder.
+TensorView* eagerMatmul(TensorView* tv_a, TensorView* tv_b) {
+  NVF_CHECK(tv_a->dtype() == tv_b->dtype());
```

Comment:
This condition is not mentioned in https://pytorch.org/docs/stable/generated/torch.matmul.html although `torch.matmul` fails with this error if you try different types:
> RuntimeError: expected mat1 and mat2 to have the same dtype, but got: c10::Half != float
Maybe we could put a comment there that this restriction might possibly be lifted in the future.